File: packages/core/src/factories/internal/metadata/iterate_metadata_escape.ts
import ts from "typescript";

import { MetadataEscaped } from "../../../schemas/metadata/MetadataEscaped";
import { MetadataSchema } from "../../../schemas/metadata/MetadataSchema";
import { Writable } from "../../../typings/Writable";
import { TypeFactory } from "../../TypeFactory";
import { IMetadataIteratorProps } from "./IMetadataIteratorProps";
import { iterate_metadata } from "./iterate_metadata";

export const iterate_metadata_escape = (
  props: IMetadataIteratorProps,
): boolean => {
  if (props.options.escape === false || props.explore.escaped === true)
    return false;

  const escaped: ts.Type | null = TypeFactory.getReturnTypeOfClassMethod({
    checker: props.checker,
    class: props.type,
    function: "toJSON",
  });
  if (escaped === null) return false;

  if (props.metadata.escaped === null) {
    Writable(props.metadata).escaped = MetadataEscaped.create({
      original: MetadataSchema.initialize(),
      returns: MetadataSchema.initialize(),
    });
  }
  iterate_metadata({
    ...props,
    metadata: props.metadata.escaped!.original,
    explore: {
      ...props.explore,
      escaped: true,
    },
  });
  iterate_metadata({
    ...props,
    metadata: props.metadata.escaped!.returns,
    type: escaped,
    explore: {
      ...props.explore,
      escaped: true,
    },
  });
  return true;
};

File: packages/core/src/factories/internal/metadata/iterate_metadata_escape.go
package metadata

func IterateMetadataEscape(props IMetadataIteratorProps) bool {
	if props.Options.Escape != nil && !*props.Options.Escape {
		return false
	}
	if props.Explore.Escaped || props.Checker == nil || props.Type == nil {
		return false
	}
	escaped := props.Checker.ReturnTypeOfMethod(props.Type, "toJSON")
	if escaped == nil {
		return false
	}
	if props.Metadata.Escaped == nil {
		props.Metadata.Escaped = &MetadataEscaped{Original: NewMetadataSchema(), Returns: NewMetadataSchema()}
	}
	IterateMetadata(IMetadataIteratorProps{Options: props.Options, Checker: props.Checker, Components: props.Components, Errors: props.Errors, Metadata: props.Metadata.Escaped.Original, Type: props.Type, Explore: props.Explore.WithEscaped(true)})
	IterateMetadata(IMetadataIteratorProps{Options: props.Options, Checker: props.Checker, Components: props.Components, Errors: props.Errors, Metadata: props.Metadata.Escaped.Returns, Type: escaped, Explore: props.Explore.WithEscaped(true)})
	return true
}
